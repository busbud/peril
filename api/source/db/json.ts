import * as JSON5 from "json5"

import { getGitHubFileContentsWithoutToken } from "../github/lib/github_helpers"
import { PERIL_ORG_INSTALLATION_ID } from "../globals"
import winston from "../logger"
import { getDB } from "./getDB"
import { DangerfileReferenceString, DatabaseAdaptor, GitHubInstallation } from "./index"

// Effectively you need to have a JSON file that looks like a GitHubInstallation,
// but also have a `repos` array of GitHubRepo -so you can do per repo rules in there.

/*
For example:
{
  "settings": {
  },
  "rules": {
   "pull_request": "orta/peril@pr.ts",
   "issue": "orta/peril@issue.ts"
  },
  "repos" : {
    "orta/ORStackView": {
      "issue.created": "orta/peril@lock_issues.ts"
    }
  }
}
*/

/** Logs */
const info = (message: string) => winston.info(`[json db] - ${message}`)
const error = (message: string) => winston.error(`[json db] - ${message}`)

let org: GitHubInstallation = null as any

export const jsonDatabase = (dangerFilePath: DangerfileReferenceString): DatabaseAdaptor => ({
  /** Gets an Integration */
  getInstallation: async (_: number): Promise<GitHubInstallation | null> => {
    if (!org) {
      const db = await getDB()
      await db.setup()
    }
    return org
  },

  /** Deletes an Integration */
  deleteInstallation: async (_: number) => {
    info(`Skipping saving github repo with slug.`)
  },

  /** Saves an Integration */
  saveInstallation: async (installation: GitHubInstallation) => {
    info(`Skipping saving installation due to no db: ${installation.iID}`)
    return installation
  },

  /** Updates the JSON for the db */
  updateInstallation: async (_: number) => {
    const db = await getDB()
    await db.setup()
    return org
  },

  /** Just always return the main one */
  getSchedulableInstallationsWithKey: () => Promise.resolve([org]),

  setup: async () => {
    const repo = dangerFilePath.split("@")[0]
    const path = dangerFilePath.split("@")[1]

    const file = await getGitHubFileContentsWithoutToken(repo, path)

    if (file === "") {
      throwNoJSONFileFound(dangerFilePath)
    }

    const parsedOrg = JSON5.parse(file) as Partial<GitHubInstallation>
    if (!parsedOrg) {
      error(`Could not run JSON.parse on the contents of ${dangerFilePath}.`)
      process.exitCode = 1
    } else {
      // Set our write-once org variable that is then re-used for all of the different
      // installation related calls
      org = partialInstallationToInstallation(parsedOrg, dangerFilePath)
    }
  },
})

/**
 * Go from a potentially semi-filled in Installation, to one
 * where you can trust that things exist.
 */
export const partialInstallationToInstallation = (
  partial: Partial<GitHubInstallation>,
  dangerfileRefString: DangerfileReferenceString
) => ({
  iID: partial.iID || PERIL_ORG_INSTALLATION_ID,
  login: partial.login || "unknown",
  avatarURL: partial.avatarURL || "", // TODO: Generate a thumbnail
  repos: partial.repos || {},
  rules: partial.rules || {},
  scheduler: partial.scheduler || {},
  tasks: partial.tasks || {},
  lambdaName: partial.lambdaName || "",
  perilSettingsJSONURL: dangerfileRefString,
  installationSlackUpdateWebhookURL: undefined,
  settings: {
    env_vars: (partial.settings && partial.settings.env_vars) || [],
    ignored_repos: (partial.settings && partial.settings.ignored_repos) || [],
    modules: (partial.settings && partial.settings.modules) || [],
    disable_github_check: (partial.settings && partial.settings.disable_github_check) || false,
  },
})

const throwNoJSONFileFound = (dangerFilePath: DangerfileReferenceString) => {
  const msg = "Could not find find a JSON file for Peril settings."
  const subtitle = `It's likely that Peril cannot connect to ${dangerFilePath}, check the logs for more info above here.`
  const action = `You'll probably need to make changes to the "DATABASE_JSON_FILE" in your ENV vars.`
  throw new Error([msg, subtitle, action].join(" "))
}
