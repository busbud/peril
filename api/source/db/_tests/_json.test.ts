const legitSettings = `{
  "id": 1,
  "settings": {
  },
  "rules": {
   "pull_request": "orta/peril@pr.ts",
   "issue": "orta/peril@issue.ts"
  },
  "repos" : {
    "orta/ORStackView": {
      "issue.created": "orta/peril@lock_issues.ts"
    }
  }
}`

const mockGHContents = jest.fn()
jest.mock("../../github/lib/github_helpers", () => ({
  getGitHubFileContentsWithoutToken: mockGHContents,
}))

import { DatabaseAdaptor } from "../index"
import { jsonDatabase } from "../json"

describe("makes the right calls to GitHub", () => {
  let db: DatabaseAdaptor = null as any

  const setup = async () => {
    mockGHContents.mockImplementationOnce(() => Promise.resolve(legitSettings))

    db = jsonDatabase("orta/peril@settings.json")
    return db.setup()
  }

  it("with a legit stubbed JSON file", async () => {
    await setup()

    const org = await db.getInstallation(1)
    expect(org).toMatchSnapshot()
  })
})

it("Raises with a bad URL", async () => {
  mockGHContents.mockImplementationOnce(() => Promise.resolve(""))
  expect.assertions(1)

  try {
    const db = jsonDatabase("orta/other@settings.json")
    await db.setup()
  } catch (error) {
    expect(error).toMatchSnapshot()
  }
})
