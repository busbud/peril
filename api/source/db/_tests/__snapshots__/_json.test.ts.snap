// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Raises with a bad URL 1`] = `[Error: Could not find find a JSON file for Peril settings. It's likely that Peril cannot connect to orta/other@settings.json, check the logs for more info above here. You'll probably need to make changes to the "DATABASE_JSON_FILE" in your ENV vars.]`;

exports[`makes the right calls to GitHub with a legit stubbed JSON file 1`] = `
Object {
  "avatarURL": "",
  "iID": 23511,
  "installationSlackUpdateWebhookURL": undefined,
  "lambdaName": "",
  "login": "unknown",
  "perilSettingsJSONURL": "orta/peril@settings.json",
  "repos": Object {
    "orta/ORStackView": Object {
      "issue.created": "orta/peril@lock_issues.ts",
    },
  },
  "rules": Object {
    "issue": "orta/peril@issue.ts",
    "pull_request": "orta/peril@pr.ts",
  },
  "scheduler": Object {},
  "settings": Object {
    "disable_github_check": false,
    "env_vars": Array [],
    "ignored_repos": Array [],
    "modules": Array [],
  },
  "tasks": Object {},
}
`;
